### ISSUE TITLE
Unnecessary constraint public_id_pid_id_UNIQUE in domain_public_ids table?

### ISSUE THREAD

**Issue by mvs3d** [NICMx/rdap-server#109]
Hello there,

Can you explain why you have constraint "public_id_pid_id_UNIQUE" in domain_public_ids definition?

With this constrant you can't make links from several domains to one publicId. 

```
-- -----------------------------------------------------
-- Table `rdap`.`domain_public_ids`
-- -----------------------------------------------------
DROP TABLE IF EXISTS `rdap`.`domain_public_ids` ;

CREATE TABLE IF NOT EXISTS `rdap`.`domain_public_ids` (
  `dom_id` BIGINT NOT NULL COMMENT 'Domain\'s id',
  `pid_id` BIGINT NOT NULL COMMENT 'Public id\'s id',
  PRIMARY KEY (`dom_id`, `pid_id`),
  INDEX `fk_domain_public_ids_public_id1_idx` (`pid_id` ASC),
  UNIQUE INDEX `public_id_pid_id_UNIQUE` (`pid_id` ASC),
  CONSTRAINT `fk_domain_public_ids_domain1`
    FOREIGN KEY (`dom_id`)
    REFERENCES `rdap`.`domain` (`dom_id`)
    ON DELETE NO ACTION
    ON UPDATE NO ACTION,
  CONSTRAINT `fk_domain_public_ids_public_id1`
    FOREIGN KEY (`pid_id`)
    REFERENCES `rdap`.`public_id` (`pid_id`)
    ON DELETE NO ACTION
    ON UPDATE NO ACTION)
ENGINE = InnoDB
COMMENT = 'This table contains the relation between a Domain and its public ids.';

```

**Comment by mvs3d** [comment]
I suppose we need unique index by dom_id here, right?

**Comment by dhfelix** [comment]
> Can you explain why you have constraint "public_id_pid_id_UNIQUE" in domain_public_ids definition?
With this constrant you can't make links from several domains to one publicId.

We assumed that publicIds are related only to one object. Couldn't find anything to prove or refuse it. Just to know more about the subject, can you share an use case where publicIds could be linked to two or more objects?

In the meantime, you could drop the UNIQUE constraint, Java Code won't crash.

> I suppose we need unique index by dom_id here, right?

no, an object could have multiples publicIds from different types.

**Comment by mvs3d** [comment]
> We assumed that publicIds are related only to one object. Couldn't find anything to prove or refuse it. Just to know more about the subject, can you share an use case where publicIds could be linked to two or more objects?

Well "domain_public_ids" table - this is table with links between domains and publicId's, no any types of object linked there, right?

Maybe I'm wrong, i just not sure where I can write publicId for domain like in this example:

https://rdap.verisign.com/com/v1/domain/logicboxes.com

There attached entity with handle "303" and role "registrar". I just make proposal I must create link between publicId (linked with enitity "303, registrar") and certain domain. Or I'm wrong?

> no, an object could have multiples publicIds from different types.

Hm.. interesting. Can you provide any links to more detailed documentation about it?

**Comment by dhfelix** [comment]
>>We assumed that publicIds are related only to one object. Couldn't find anything to prove or refuse it. Just to know more about the subject, can you share an use case where publicIds could be linked to two or more objects?
> Well "domain_public_ids" table - this is table with links between domains and publicId's, no any types of object linked there, right?

domain_public_ids, the link is only between domain and publicId.
entity_public_ids, the link is only betwork entity and publicId.

When I said object, I meant both domains and entities.. You could check the E-R diagram here https://www.reddog.mx/img/diagram/reddog-db-er.svg

> Maybe I'm wrong, i just not sure where I can write publicId for domain like in this example:
https://rdap.verisign.com/com/v1/domain/logicboxes.com

Oh!, the publicID is related to the entity not the domain in that example. e.g.
```
{
    "objectClassName": "domain",
    "handle": "97845829_DOMAIN_COM-VRSN",
    "ldhName": "LOGICBOXES.COM",
   .
   .
   "entities": [
        {
            "objectClassName": "entity",
            "handle": "303",
            "roles": [
                "registrar"
            ],
            "publicIds": [
                {
                    "type": "IANA Registrar ID",
                    "identifier": "303"
                }
            ],
         ...
         ...
    }]
.....
....
}
```

If you write it directly to "domain_public_ids" table, the response will be like 
```
{
    "objectClassName": "domain",
    "handle": "97845829_DOMAIN_COM-VRSN",
    "ldhName": "LOGICBOXES.COM",
   "publicIds": [
                {
                    "type": "IANA Registrar ID",
                    "identifier": "303"
                }
            ],
    "entities": [
           {
               "objectClassName": "entity",
               "handle": "303",
               "roles": [
                   "registrar"
                ],
         ...
         ...
    }]
.....
....
}
}
```

You could achieve the first response by creating an entity, then link it to its own publicId , finally N domains can be linked to that entity with a role "registrar" (domain_entity_roles table).

>> no, an object could have multiples publicIds from different types.
> Hm.. interesting. Can you provide any links to more detailed documentation about it?

https://tools.ietf.org/html/rfc7483#section-4.8 publicID definition for RDAP

**Comment by mvs3d** [comment]
> domain_public_ids, the link is only between domain and publicId.
> entity_public_ids, the link is only betwork entity and publicId.
> 
> When I said object, I meant both domains and entities.. You could check the E-R diagram here https://www.reddog.mx/img/diagram/reddog-db-er.svg

Yep, I understand this. 

> You could achieve the first response by creating an entity, then link it to its own publicId , finally N domains can be linked to that entity with a role "registrar" (domain_entity_roles table).

Ah.. OK, clearly for me now. Thanks :)

> https://tools.ietf.org/html/rfc7483#section-4.8 publicID definition for RDAP

Well, there an example for using public for Entity (Registrar) only. What I don't understand, in which cases we need use domain_public_ids table? And still very strange looks for me UNIQUE constrant by publicId in this, relation table. This mean we can have only ONE domain related with ONE publicId. This is looks good for entities, but not for domains.

**Comment by dhfelix** [comment]
> Well, there an example for using public for Entity (Registrar) only. What I don't understand, in which cases we need use domain_public_ids table?

RFC7483#Section5.3 is the definition of domain object class for rdap response, there the class members are listed.
publicIds included and only makes a reference to section 4.8.

Figure 23 shows an example of a JSON domain object representing a forward DNS delegation point that might be served by a DNR, and that example includes a publicId in domain object.

```
 {
     "objectClassName" : "domain",
     "handle" : "XXXX",
     "ldhName" : "xn--fo-5ja.example",
     "unicodeName" : "foo.example",
     .
     .
     .
     "publicIds":[
       {
         "type":"ENS_Auth ID",
         "identifier":"1234567890"
       }
     ],
     .
     .
}
```

Eligibility and Name-Selection Services (ENS Services), is used as publicId, but honestly, I know nothing about it, and didn't find info.

I was hoping that you know more about it.

> And still very strange looks for me UNIQUE constrant by publicId in this, relation table. This mean we can have only ONE domain related with ONE publicId. This is looks good for entities, but not for domains.

you could drop the UNIQUE constraint, or if you have information to support this drop, we can make it official.

**Comment by mvs3d** [comment]
> Eligibility and Name-Selection Services (ENS Services), is used as publicId, but honestly, I know nothing about it, and didn't find info.
> 
> I was hoping that you know more about it.

Hm.. no, I haven't any info about it. I just can propose we can use publicId as another (like dom_handle) unique identifier for every domain.

> you could drop the UNIQUE constraint, or if you have information to support this drop, we can make it official.

Well, at current time I schedule just don't use publicIDs for domains, because can't find full information about it. I suppose we can close this issue.
